orchestra/utils/revert.py

```diff
@@ -0,0 +1,258 @@
+
+    Args:
+        task (orchestra.models.Task):
+            The ID of the task to be reverted.
+        iteration_id (datetime.datetime):
+            The ID of the iteration before which to revert the task.
+        revert_before (bool):
+            If `revert_before` is indicated, the task will be reverted to a
+            state before the specified iteration was created; otherwise, it
+            will be reverted to a state in which the iteration is processing.
+        commit (bool):
+            Determines whether the revert is actually carried out;
+            otherwise, an audit trail is passively generated.
+
+    Returns:
+        audit (dict):
+            An audit trail for the revert, e.g.,
+            {
+                'task': {...},
+                'change': 1
+                'assignments': [
+                    {
+                        # worker_0 assignment
+                        'assignment': {...},
+                        'change': 1,
+                        'iterations': [
+                            'iterations': [
+                                {
+                                    'iteration': {...},
+                                    'change': 2
+                                }
+                            ]
+                        ]
+                    },
+                    {
+                        # worker_1 assignment
+                        'assignment': {...},
+                        'change': 2,
+                        'iterations': []
+                    }
+                ],
+            }
+    """
+    task = Task.objects.get(id=task_id)
+    revert_iteration = Iteration.objects.get(id=iteration_id)
+    task_audit = _build_revert_audit(task, revert_iteration, revert_before)
+    if commit:
+        _revert_task_from_audit(task_audit)
+    return task_audit
+
+
+def _revert_task_from_audit(task_audit):
+    task = Task.objects.get(id=task_audit['task']['id'])
+    if task_audit['change'] == RevertChange.DELETED.value:
+        task.delete()
+    elif task_audit['change'] == RevertChange.REVERTED.value:
+        task.status = task_audit['reverted_status']
+        task.save()
+        for assignment_audit in task_audit['assignments']:
+            _revert_assignment_from_audit(assignment_audit)
+
+
+def _revert_assignment_from_audit(assignment_audit):
+    assignment = TaskAssignment.objects.get(
+        id=assignment_audit['assignment']['id'])
+    if assignment_audit['change'] == RevertChange.DELETED.value:
+        assignment.delete()
+    elif assignment_audit['change'] == RevertChange.REVERTED.value:
+        for iteration_audit in assignment_audit['iterations']:
+            _revert_iteration_from_audit(iteration_audit)
+
+        assignment.refresh_from_db()
+        if get_latest_iteration(
+                assignment.id).status == Iteration.Status.PROCESSING:
+            assignment.status = TaskAssignment.Status.PROCESSING
```

Comment:
leave a TODO for when we archive data to set the latest data to the last submitted data